### ISSUE TITLE
iCub eyes version/vergence coupling handlers

### ISSUE THREAD

**Issue by xEnVrE** [robotology/gazebo-yarp-plugins#473]
I am trying to understand the code of the coupling handlers for the eyes.

### From physical joints space to controlled DoFs space

`decouplePos` should move from the physical joint space (i.e. `left` and `right` eyes) to the controlled DoFs space (i.e.`eyes-version` and `eyes-vergence`).

https://github.com/robotology/gazebo-yarp-plugins/blob/8c056e21c10bd231d1659612285386f12432626b/plugins/controlboard/src/ControlBoardDriverCoupling.cpp#L77-L84

Then, shouldn't it be something like this?

```
// version = (left + right)  / 2.0
// vergence = left - right
 double left = current_pos[m_coupledJoints[0]]; 
 double right =  current_pos[m_coupledJoints[1]]; 
 current_pos[m_coupledJoints[0]] = (left + right) / 2.0; 
 current_pos[m_coupledJoints[1]] = left - right; 
```

### From controlled DoFs space to physical joints space

`decoupleRefPos` should move from the controlled DoFs space (i.e.`eyes-version` and `eyes-vergence`) to the physical joint space (i.e. `left` and `right` eyes)

https://github.com/robotology/gazebo-yarp-plugins/blob/8c056e21c10bd231d1659612285386f12432626b/plugins/controlboard/src/ControlBoardDriverCoupling.cpp#L110-L117

Then, shouldn't it be something like this?

```
// left   = version + vergence / 2.0
// right = version - vergence / 2.0
out[m_coupledJoints[0]] = pos_ref[m_coupledJoints[0]] + pos_ref[m_coupledJoints[1]]/2; 
out[m_coupledJoints[1]] = pos_ref[m_coupledJoints[0]] - pos_ref[m_coupledJoints[1]]/2; 
```

### Summary

To summarize, the plugin is implementing these relationships

```
version = left + right
vergence = left - right

left = (version + vergence)/2
right = (version - vergence)/2
```

while, according to this [documentation](http://wiki.icub.org/wiki/Vergence,_Version_and_Disparity), I was expecting
```
version = (left + right) / 2
vergence = left - right

left = version + (vergence / 2)
right = version - (vergence / 2)
```

**Comment by xEnVrE** [comment]
cc @traversaro

**Comment by traversaro** [comment]
Thanks a lot @xEnVrE for the clear explanation. 

I wonder if @randaz81 or @aerydna remember anything related to this, however it may be tricky because the development dates back to  approximately four years ago. @pattacini may also want to weigh in on this, clearly with absolutely low priority when he has time given the current situation. 

In any case, I try to summarize what you report in a more compact way (correct me if I am wrong): 

##  Eyes

It seems that  gazebo-yarp-plugins' `EyesCouplingHandler` defines the version (in the sense that with definition the existing`decouplePos` and `decoupleRefPos` are coherent) as : 
~~~
version = left + right
~~~
instead of the definition used for  example in https://github.com/robotology/QA/issues/158 and http://wiki.icub.org/wiki/Vergence,_Version_and_Disparity of: 
~~~
version = (left + right)/2
~~~

## Finger adduction/abduction

The  situation is less clear to me here. Probably it would be useful to have a  clear definition with names of the physical and controlled  joints, see https://github.com/robotology/icub-models/issues/30 . If you have a proposal feel free to drop it there.

**Comment by traversaro** [comment]
cc @kouroshD

**Comment by pattacini** [comment]
I can confirm that the law implemented in the gaze controller's code is:
```
version = (left + right)/2
```
as per our documentation.
I'm also reasonably sure that the FW copies that as well.

**Comment by traversaro** [comment]
I dug a bit in old  issues, and I found the original issue related to eyes support in https://github.com/robotology/gazebo-yarp-plugins/issues/149#issue-47657172 . In that issue, apparently I am the one that reported the definition that is currently implemented, without the division by two: 
~~~
eyes_version    =  left_eye_pan + right_eye_pan
~~~
Even if I was  reporting the the Wiki page that contained the `version = (left + right)/2`, I apparently disregarded it on the basic of what was actually implemented in the firmware, i.e. quoting the original issue: 
> as you can check directly in the iCub firmware source code in firmware/motorControllerDsp56f807/position_decoupling.c, line 25 or in firmware/motorControllerDsp56f807/pwm_decoupling.c, line 83

Interestingly, the relative line of code in the firmware have not been modified in the meanwhile, see: 
* https://github.com/robotology/icub-firmware/blob/fcace9ed15ad99b3a8da656d1782ab3a19c7f7f9/motorControllerDsp56f807/common_source_code/position_decoupling.c#L25
* https://github.com/robotology/icub-firmware/blob/fcace9ed15ad99b3a8da656d1782ab3a19c7f7f9/motorControllerDsp56f807/common_source_code/pwm_decoupling.c#L82

In particular, the "version" numbers are documented in https://github.com/robotology/icub-firmware/blob/fcace9ed15ad99b3a8da656d1782ab3a19c7f7f9/motorControllerDsp56f807/4DC/source_code/include/options.h#L9 , and the  one related to the eyes (0x0215 and 0x0115) seems both to use the  formula `eyes_version    =  left_eye_pan + right_eye_pan` . So unless I  am missing something in the firmware (that is perfectly possible) it  seems that the  gazebo-yarp-plugins' EyesCouplingHandler is coherent with the iCub firmware behaviour. 

. 

PS: @xEnVrE given the complexity of the issue, probably  it could make sense to move the discussion on the finger abduction to a separate issue to simplify discussions.

**Comment by xEnVrE** [comment]
> PS: @xEnVrE given the complexity of the issue, probably it could make sense to move the discussion on the finger abduction to a separate issue to simplify discussions.

Yep!

**Comment by pattacini** [comment]
Eh! If this is true, that's a very good catch then @traversaro 
I've never browsed the FW myself but I clearly remember discussions that gravitated around our main Wiki resource touching on the requirements on the FW as well.

Anyway, let me tone down my assertion above, I can confirm that the gaze controller implements the law with the 2. Now wondering what could be happening with this mismatch between the high-level and the low-level policies. Perhaps the closed-loop is saving our day?

**Comment by xEnVrE** [comment]
I updated the first comment of the issue, so that my doubts are explained more clearly.

**Comment by traversaro** [comment]
I think I  also found the decoupling code also for ETH-based boards, but also on that case the coupling assumed is `eyes_version    =  left_eye_pan + right_eye_pan`, see: 
https://github.com/robotology/icub-firmware/blob/v1.15.0/emBODY/eBcode/arch-arm/embobj/plus/mc/Controller.c#L1127 .

**Comment by traversaro** [comment]
Ok, last stage: I think the coupling information is now stored in configuration files, but even there the assumed version coupling seems to be `eyes_version = left_eye_pan + right_eye_pan` : 
https://github.com/robotology/robots-configuration/blob/v1.15.0/iCubGenova02/hardware/mechanicals/head-eb21-j2_5-mec.xml#L34

**Comment by xEnVrE** [comment]
Can this be closed?

**Comment by traversaro** [comment]
If they work fine, ok for me!
